react/chakra-ui (#75)
infinum/frontend-handbook: React/CSS-in-JS/Chakra.md

# Chakra UI

Chakra UI is a simple, modular and accessible component library that gives you the building blocks you need to build your React applications.

## Getting started

### Installation:

```
npm i @chakra-ui/react @emotion/react @emotion/styled framer-motion
```

### Setup provider:

> For Chakra UI to work correctly, you need to setup the `ChakraProvider` at the root of your application.

Reviewer: `setup` is a noun, use `set up` as a verb
```suggestion
For Chakra UI to work correctly, you need to set up the `ChakraProvider` at the root of your application.
```

Example:

```jsx
import { ChakraProvider } from "@chakra-ui/react";

function App({ Component, pageProps }: AppProps): ReactElement {
  return (
    <ChakraProvider theme={theme}>
      <Component {...pageProps} />
    </ChakraProvider>
  );
}

export default App;
```

Note:

- For Next.js, you need to set this up in `pages/_app.tsx`
- For Create React App, you need to set this up in `index.tsx`

> 

Reviewer 2: Add a note that ChakraProvider uses emotion ThemeProvider under the hood. 
https://github.com/chakra-ui/chakra-ui/blob/develop/packages/react/src/chakra-provider.tsx#L57
https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/providers.tsx#L12

Mention that it uses [styled-system](https://styled-system.com/) too
https://github.com/chakra-ui/chakra-ui/blob/develop/packages/styled-system/src/system.ts#L1
Explain [responsive-styles](https://styled-system.com/responsive-styles) props 

Pseudo props should also be mentioned somewhere https://chakra-ui.com/docs/features/style-props#pseudo, this is specific to Chakra UI

### Style props

Style props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components.

Alternative to that is using sx prop that takes an object with style rules.

```tsx
// shorthand width prop
<Box w="400px">Lorem Ipsum is simply dummy text</Box>

// sx prop
<Box sx={{ width: '400px'}}>Lorem Ipsum is simply dummy text</Box>
```

For more about style props read [docs](https://chakra-ui.com/docs/features/style-props)

> #### The `as` prop

Reviewer 2: This is called `polymorphic prop` and it's actually a feature of Emotion https://emotion.sh/docs/styled#as-prop borrowed from styled-components https://styled-components.com/docs/api#as-polymorphic-prop.
You should mention it in this section.

The `as` prop is a feature that all Chakra UI components have and it allows you to pass an HTML tag or component to be rendered.

> For example, say you are using a Button component, and you need to make it a link instead. You can compose a and Button like this:

Reviewer: To make it clearer that `a` is not the article from English language (a/the), but the `anchor` element:

```suggestion
For example, say you are using a `Button` component, and you need to make it a link instead. You can compose `a` and `Button` like this:
```

```tsx
<Button as="a" variant="outline" href="https://chakra-ui.com">
  Lorem Ipsum
</Button>
```

> This allows you to use all of the Button props and all of the a props without having to wrap the Button in an a component.

Reviewer: ```suggestion
This allows you to use all of the `Button` props and all of the `a` props without having to wrap the `Button` in an `a` component.
```

> This allows you to use all of the `Button` props and all of the `a` props without having to wrap the `Button` in an `a` component.

Reviewer 2: You can mention that this also work `<Button as={Link} someLinkProp={value} />`
- Link is another component
- Link props will become available on Button
- styles will be combined

Example:

```jsx
import { extendTheme } from "@chakra-ui/react";

const overrides = {
  colors: {
    primary: {
      100: "#ff0000",
      80: "#ff1a1a",
    },
  },
};

const theme = extendTheme(overrides);

export default theme;
```

For more info about setup read [docs](https://chakra-ui.com/docs/getting-started)

## Theming rules

The theme object is where you define your application's color palette, type scale, font stacks, breakpoints, border radius values, and more. Theme overrides are placed in `src/styles/themes` folder.

Example:

```ts
const overrides = {
  colors: {
    primary: {
      100: "#ff0000",
      80: "#ff1a1a",
    },
  },
};
```

Except objects, theme override can take a function:

```ts
const overrides = {
  colors: (props) => {
    const { colorMode } = props;

    return {
      primary: {
        100: colorMode === "dark" ? "#ff0000" : "#ffffff",
        80: "#ff1a1a",
      },
    };
  },
};
```

Then in `props` we have access button props, `theme` object and `colorMode`.

Note:

- `mode(lightMode, darkMode)(props)` function is the same as `colorMode === "dark" ? darkMode : lightMode`.

> - `mode` function is part of `chakra-ui/theme-tools` package.

Reviewer 2: there is also `orient` https://github.com/chakra-ui/chakra-ui/blob/develop/packages/theme-tools/src/component.ts#L44 
Example: https://github.com/chakra-ui/chakra-ui/blob/e20efc1d8d42ce9d3c2280a5731d59e771de53c4/packages/theme/src/components/slider.ts#L8

And a lot of color transformation helpers https://github.com/chakra-ui/chakra-ui/blob/develop/packages/theme-tools/src/color.ts
`transparentize` is a cool one
Example: https://github.com/chakra-ui/chakra-ui/blob/abac2f999f3996ffc3ce575de07126dc9cc41bbb/packages/theme/src/components/badge.ts#L13

      },
    },
  },
};
```

### Responsive

Chakra UI supports responsive styles out of the box. Instead of manually adding @media queries and adding nested styles throughout your code, Chakra UI allows you to provide object and array values to add mobile-first responsive styles.

> Under the hood `@media(min-width)` media query is used to ensure mobile-first

Chakra UI default breakpoints:

```ts
export const breakpoints = {
  sm: "30em",
  md: "48em",
  lg: "62em",
  xl: "80em",
};
```

Here's how to interpret this syntax:

- `base`: From `0em` upwards
- `md`: From `48em` upwards
- `lg`: From `62em` upwards
- `xl`: From `80em` upwards

For responsive styles, array or object syntax can be used.

```tsx
// array syntax
<Text fontSize={["24px", "40px", "56px"]}>
  Lorem Ipsum is simply dummy text
</Text>
```

```tsx
// object syntax
<Text fontSize={{ base: "24px", md: "40px", lg: "56px" }}>
  Lorem Ipsum is simply dummy text
</Text>
```

In case we need to skip cretan breakpoint, `null` is passed at that position in the array to avoid generating unnecessary CSS.

Example:

```jsx
<Text fontSize={["24px", null, "56px"]}>Lorem Ipsum is simply dummy text</Text>
```

> Array and Object syntax work for every style prop in the them specification, which means you can change most properties at a given breakpoint.

Reviewer: What did you mean by `in the them specification`?

Author: theme*

> To create custom breakpoints read [docs](https://chakra-ui.com/docs/features/responsive-styles#customizing-breakpoints).

Reviewer 2: Maybe describe `orient` here also.

### Custom components style

Chakra UI has a specific API for styling components. Most components have default or base styles `baseStyle`, styles for different sizes `sizes`, and styles for different visual variants `variants`.

Components styles are defined in `theme/components/<component-name>.ts`
Each component style will export these objects:

- `baseStyles`
- `sizes`
- `variants`
- `defaultProps`

Base style are styles that all button types share.

```tsx

> const baseStyle = {

Reviewer: How would you ensure that there are no typos in this object? For example: `lineHeigth` instead of `lineHeight`? Is there a type for this object?

Also how would you ensure that only valid values are used? For example: `cursor: "foo"`

Author: > How would you ensure that there are no typos in this object? For example: lineHeigth instead of lineHeight? Is there a type for this object?
I added `StyleObjectOrFn` type for that

> Also how would you ensure that only valid values are used? For example: cursor: "foo"
This is a problem because chakra doesn't have a type for that and it would be hard to create one because the value can be some css value or something from a theme or array (cc @isBatak)

  lineHeight: "1.2",
  borderRadius: "md",
  _focus: {
    boxShadow: "outline",
  },
  _disabled: {
    opacity: 0.4,
    cursor: "not-allowed",
    boxShadow: "none",
  },
  _hover: {
    cursor: "pointer",
  },
};
```

<!-- TODO change this example to use color schema -->

Variant represents visual style.

```tsx
const variants = {
  solid: {
    bg: 'primary.100',,
    color: 'white',
    _hover: {
      bg: 'primary.400',
    },
    _disabled: {
      bg: 'neutral.200',
    }
  },
	outline: {
    bg: 'transparent',,
    border: '1px solid',
    borderColor: 'primary.100',
    _hover: {
      bg: 'primary.100',
    }
  },
};
```

> Sizes determent all the component sizes like `width`, `height`, `font-size` etc

Reviewer: Did you mean `determine`?

```tsx
const sizes = {
  sm: {
    h: 8,
    minW: 8,
    fontSize: "sm",
    px: 3,
  },
  xs: {
    h: 6,
    minW: 6,
    fontSize: "xs",
    px: 2,
  },
};
```

Naming used for sizes:

- `lg`
- `md`
- `sm`
- `xs`

> Default values are default values for `size` and `variant`

Reviewer: This can be a bit confusing to read, maybe change it to this?
```suggestion
`defaultProps` are default values for `size` and `variant`
```

Similar comment for `Sizes determent all` as well, maybe change it to
```
`sizes` determine....
```

### Color mode

When you use the ChakraProvider at the root of your app, you can automatically use color mode in your apps. By default, most of Chakra UI component are dark mode compatible. To handle color mode manually in your application, use the `useColorMode` or `useColorModeValue` hooks.

> Tip: Chakra stores the color mode in localStorage and uses CSS variables to ensure the color mode is persistent.

#### useColorMode

`useColorMode` is a React hook that gives you access to the current color mode, and a function to toggle the color mode.

```tsx
function Example() {
  const { colorMode, toggleColorMode } = useColorMode();
  return (
    <header>
      <Button onClick={toggleColorMode}>
        Toggle {colorMode === "light" ? "Dark" : "Light"}
      </Button>
    </header>
  );
}
```

#### useColorModeValue

`useColorModeValue` is React hook that takes 2 arguments, first is value for light mode and second is value for dark mode, and returnees the the value based on the active color mode

```tsx
const value = useColorModeValue(lightModeValue, darkModeValue);
```

> 

Reviewer 2: Maybe we could add one more section dedicated to the chakra factory https://chakra-ui.com/docs/features/chakra-factory.
It could be used similar to emotion `styled` component but it will be aware of the theme props.
```
const Card = chakra("div", {
  baseStyle: {
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  },
})
```

Author: I added chakra factory section + example with chakra factory function and `react-datepicker`, similar to there's `react-input-autoresize` [example](https://chakra-ui.com/docs/features/chakra-factory#chakra-factory-function)

- clearIndicator
- container
- control
- dropdownIndicator
- group
- groupHeading
- indicatorsContainer
- indicatorSeparator
- input
- loadingIndicator
- loadingMessage
- menu
- menuList
- menuPortal
- multiValue
- multiValueLabel
- multiValueRemove
- noOptionsMessage
- option
- placeholder
- singleValue
- valueContainer

Reade more about react-select style objects in [docs](https://react-select.com/styles#style-object).

After we have `ReactSelect` theme object, we need to add that to `theme.components`

```tsx
// style/theme/index.ts
const overrides = {
  colors,
  // ...
  components: {
    Button,
    ReactSelect,
    // ...
  },
};
```

Now we can create Select component in components folder:

```tsx
const selectStyle = {
  container: (base, { theme }) => {
    return {
      ...base,
      ...theme.container,
    };
  },
  menu: (base, { theme }) => ({
    ...base,
    ...theme.menu,
  }),
};

const Select = (props) => {
  const styles = useStyleConfig("ReactSelect", props);

  return (
    <ReactSelect
      theme={css(styles)(useTheme())}
      styles={selectStyle}
      {...props}
    />
  );
};
```

Each key added to the theme needs to be added to `selectStyle`.

> Additionally custom components can be added to react-select like `SelectContainer`

Reviewer: ```suggestion
Additionally, custom components can be added to react-select like `SelectContainer`
```

# Chakra UI

Chakra UI is a simple, modular and accessible component library that gives you the building blocks you need to build your React applications.

## Getting started

### Installation:

```
npm i @chakra-ui/react @emotion/react @emotion/styled framer-motion
```

For more about setup read [docs](https://chakra-ui.com/docs/getting-started)

### Setup provider:

For Chakra UI to work correctly, you need to set up the `ChakraProvider` at the root of your application.
Under the hood Chakra UI is using [emotion ThemeProvider](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/providers.tsx)

Example:

```jsx
import { ChakraProvider } from "@chakra-ui/react";

function App({ Component, pageProps }: AppProps): ReactElement {
  return (
    <ChakraProvider theme={theme}>
      <Component {...pageProps} />
    </ChakraProvider>
  );
}

export default App;
```

Note:

- For Next.js, you need to set this up in `pages/_app.tsx`
- For Create React App, you need to set this up in `index.tsx`

### Style props

> Style props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components.

Reviewer 3: I guess this would be preffered way of doing it?

> Alternative to that is using sx prop that takes an object with style rules.

Reviewer 4: It's fine to mention alternatives, but the preferred way of using it should be clear, together with some examples when the alternative might be better.

Reviewer 2: Style props for one-off styles
```
<Button display="block" mt="md" />
```

`sx` prop when you have to merge multiple objects
```
const ExoticComponent = (props) => {
  const styles = useStyleConfig('ExoticComponent', props);
 
  const defaultStyles = {
   display: 'block',
  };

  return <chakra.div sx={{ ...styles, ...defaultStyles }} />;
}
```
You have control over specificity and merge order. 
Also, there is `__css` which will get merged first, before the `sx`, so I guess we should use `__css` in this case.
https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L106

```
/**
 * Style resolver function that manages how style props are merged
 * in combination with other possible ways of defining styles.
 *
 * For example, take a component defined this way:
 * ```jsx
 * <Box fontSize="24px" sx={{ fontSize: "40px" }}></Box>
 * ```
 *
 * We want to manage the priority of the styles properly to prevent unwanted
 * behaviors. Right now, the `sx` prop has the highest priority so the resolved
 * fontSize will be `40px`
 */
 ```

```tsx
// shorthand width prop
<Box w="400px">Lorem Ipsum is simply dummy text</Box>

// sx prop
<Box sx={{ width: '400px'}}>Lorem Ipsum is simply dummy text</Box>
```

For more about style props read [docs](https://chakra-ui.com/docs/features/style-props)

#### The `as` polymorphic prop

The `as` prop is a feature that all Chakra UI components have and it allows you to pass an HTML tag or component to be rendered.

> `as` prop (`polymorphic prop`) is feature from emotion borrowed from styled-components

Reviewer 4: ```suggestion
`as` prop (`polymorphic prop`) is a feature of emotion borrowed from styled-components
```

### Colors

Color naming should a single value or an object with keys in range from 50 to 900.

Example:

```ts
export const colors = {
  white: "#ffffff",
  primary: {
    50: "#f6faff",
    100: "#d7e9fd",
    200: "#c0dcfc",
    300: "#a7cefb",
    400: "#8abdfa",
    500: "#68aaf8",
    600: "#3d92f6",
    700: "#0070f3",
    800: "#0064d8",
    900: "#0055b9",
  },
};
```

### Pseudo props

Pseudo props in ChakraUI can be passed as props to component. Full list of props can be found in [docs](https://chakra-ui.com/docs/features/style-props#pseudo)

Example:

```tsx
<Button
  bg="blue.100"
  _hover={{
    background: "white",
  }}
>
  Submit
</Button>
```

Also pseudo elements can be used in theme or in `sx` prop

```tsx
<Button
  sx={{
    bg: 'blue.100',
    _hover: {
      bg: 'white
    }
  }}
>
  Submit
</Button>

// or

<Button
  sx={{
    bg: 'blue.100',
    ':hover': {
      bg: 'white
    }
  }}
>
  Submit
</Button>
```

### Global styles

Global styles are theme-aware styles you can apply to any html element globally.

They are defined in `src/styles/theme/styles.ts` file.

<!-- TODO maybe what should be defined in global -->

Example:

> 

Reviewer 3: Can we also mention `resetCss` prop on ChakraProvider?

https://chakra-ui.com/docs/migration#2-update-the-themeprovider

Author: I will add it

```ts
const overrides = {
  styles: {
    global: {
      body: {
        fontFamily: "body",
        bg: "white",
        color: "primary.200",
      },
      "*": {
        boxSizing: "border-box",
      },
    },
  },
};
```

### Responsive

Chakra UI supports responsive styles out of the box. Instead of manually adding @media queries and adding nested styles throughout your code, Chakra UI allows you to provide object and array values to add mobile-first responsive styles.

> Under the hood `@media(min-width)` media query is used to ensure mobile-first

Chakra UI default breakpoints:

```ts
export const breakpoints = {
  sm: "30em",
  md: "48em",
  lg: "62em",
  xl: "80em",
};
```

Here's how to interpret this syntax:

- `base`: From `0em` upwards
- `md`: From `48em` upwards
- `lg`: From `62em` upwards
- `xl`: From `80em` upwards

For responsive styles, array or object syntax can be used.

```tsx
// array syntax
<Text fontSize={["24px", "40px", "56px"]}>
  Lorem Ipsum is simply dummy text
</Text>
```

```tsx
// object syntax
<Text fontSize={{ base: "24px", md: "40px", lg: "56px" }}>
  Lorem Ipsum is simply dummy text
</Text>
```

> In case we need to skip cretan breakpoint, `null` is passed at that position in the array to avoid generating unnecessary CSS.

Reviewer 4: ```suggestion
In case we need to skip a certain breakpoint, `null` is passed at that position in the array to avoid generating unnecessary CSS.
```

Example:

```jsx
<Text fontSize={["24px", null, "56px"]}>Lorem Ipsum is simply dummy text</Text>
```

> Array and Object syntax work for every style prop in the theme specification, which means you can change most properties at a given breakpoint.

Reviewer 4: Again, we should define which one is the preferred way of using it.

For more about responsive styles read [styled-system responsive styles docs](https://styled-system.com/responsive-styles/)

### Chakra Factory

Chakra factory serves as an object of chakra JSX elements, and also a function that can be used to enable custom component receive chakra's style props.

```tsx
import { chakra } from "@chakra-ui/react";

// object
<chakra.button
  sx={{
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  }}
>
  Click me
</chakra.button>;

// function
chakra("button", {
  baseStyle: {
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  },
});
```

This reduces the need to create custom component wrappers and name them. Syntax is available for common html elements. See the reference for the full [list of elements](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/system.utils.ts#L9) supported.

For example `react-datepicker` can be wrapped in chakra factory function so we can pass `sx` or style props to `<DatePicker />`

Example:

```tsx
import DatePicker from "react-datepicker";
import { chakra } from "@chakra-ui/react";

> const ChakraDatepicker = chakra(DatePicker);

Reviewer 2: Can we use `Styled` prefix... Chakra use it in the source code... so
const StyledaDatepicker = chakra(DatePicker);

Reviewer 2: For example here https://github.com/chakra-ui/chakra-ui/blob/develop/packages/menu/src/menu.tsx#L71

export const Datepicker = () => (
  <ChakraDatepicker sx={{ width: "100%", bg: "blue.100" }} />
);
```

Chakra factory function will pass `className` to `DatePicker input` component with all styles. In this case `input` will be rendered with custom `bg` and `width` styles.

For more about chakra factory read [docs](https://chakra-ui.com/docs/features/chakra-factory)

### Custom components style

Chakra UI has a specific API for styling components. Most components have default or base styles `baseStyle`, styles for different sizes `sizes`, and styles for different visual variants `variants`.

Components styles are defined in `theme/components/<component-name>.ts`
Each component style will export these objects:

- `baseStyles`
- `sizes`
- `variants`
- `defaultProps`

`baseStyle` are styles that all button types share.

```tsx
const baseStyle: StyleObjectOrFn = {
  lineHeight: "1.2",
  borderRadius: "md",
  _focus: {
    boxShadow: "outline",
  },
  _disabled: {
    opacity: 0.4,
    cursor: "not-allowed",
    boxShadow: "none",
  },
  _hover: {
    cursor: "pointer",
  },
};
```

> <!-- TODO change this example to use color schema -->

Reviewer 4: um?

`variants` represents visual style.

```tsx
const variants: { [variant: string]: StyleObjectOrFn } = {
  solid: {
    bg: 'primary.100',,
    color: 'white',
    _hover: {
      bg: 'primary.400',
    },
    _disabled: {
      bg: 'neutral.200',
    }
  },

> 	outline: (props: Dict): SystemStyleObject => ({

Reviewer 4: Seems like you mixed spaces and tabs here :)

```tsx
const ListWrapper = chakra(Box, {
  baseStyle: {
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  },
});
```

For more about style props read [docs](https://chakra-ui.com/docs/features/style-props)

`sx` prop works similar to `style` prop. It takes an object with styles.

Example:

```tsx
<Button
  sx={{
    display: "block",
    mt: "md",
    color: "primary.100",
    mx: 12,
  }}
>
  Click me!
</Button>
```

There is also `__css` prop that is similar to `sx`. The only difference is that `__css` prop will be merged before `sx` prop ([github](https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L106)).

### Chakra Factory

Chakra factory serves as an object of chakra JSX elements, and also a function that can be used to enable custom component receive chakra's style props.

```tsx
import { chakra } from "@chakra-ui/react";

// preferred way!
chakra("button", {
  baseStyle: {
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  },
});

// using sx prop
<chakra.button
  sx={{
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  }}
>
  Click me
</chakra.button>;
```

> This reduces the need to create custom component wrappers and name them. Syntax is available for common html elements. See the reference for the full [list of elements](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/system.utils.ts#L9) supported.

Reviewer 5: The list just contains elements that you can put in as a string.
You can easily pass any element to chakra:

```
chakra((props) => <someelement {...props} />, {...})
```

Or since `className` is how chakra styles components:

```
chakra(({className}) => <someelement {...{className}} />, {...})
```

> Chakra factory function should be only used in atom components. They will be defined with base style and can be overridden by `sx` prop because `baseStyle` is merged before `sx` ([merge order](https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L104))

Reviewer 5: I don't think we should limit chakra factory to be used only in atom components. They should be allowed (and preferred) everywhere except pages (as we discussed).

Components created using chakra factory can be overridden by also using chakra factory (which would in my opinion also be the preferred way), so let's mention that here.

```
const Button = chakra("button", {...});
const FbButton = chakra(Button, {...});
```

#### Custom component example

For example `react-datepicker` can be wrapped in chakra factory function so we can pass `sx` or style props to `<DatePicker />`

Example:

```tsx
import DatePicker from "react-datepicker";
import { chakra } from "@chakra-ui/react";

const StyledaDatepicker = chakra(DatePicker);

> export const Datepicker = () => <StyledaDatepicker w="100%" bg="blue.100" />;

Reviewer 5: Well, you're already using the chakra factory so might as well style it there.

```
const Datepicker = chakra(DatePicker, {
  baseStyle: {
    w: '100%', // or 'full'
    bg: 'blue.100',
  },
});
```

Reviewer 2: It seems this demonstrates how to use theme-aware one-off styles with components that are not part of chakra.
A SImilar thing could be achieved with polymorphic `as` prop.
`<Box as="DatePicker" w="100%" bg="blue.100" />`
This is totally fine to me...

> Under the hood `@media(min-width)` media query is used to ensure mobile-first

Chakra UI default breakpoints:

```ts
export const breakpoints = {
  sm: "30em",
  md: "48em",
  lg: "62em",
  xl: "80em",
};
```

Here's how to interpret this syntax:

- `base`: From `0em` upwards
- `md`: From `48em` upwards
- `lg`: From `62em` upwards
- `xl`: From `80em` upwards

For responsive styles, array or object syntax can be used.

```tsx
// array syntax
<Text fontSize={["24px", "40px", "56px"]}>
  Lorem Ipsum is simply dummy text
</Text>
```

```tsx
// object syntax
<Text fontSize={{ base: "24px", md: "40px", lg: "56px" }}>
  Lorem Ipsum is simply dummy text
</Text>
```

In case we need to skip a certain breakpoint, `null` is passed at that position in the array to avoid generating unnecessary CSS.

Example:

```jsx
<Text fontSize={["24px", null, "56px"]}>Lorem Ipsum is simply dummy text</Text>
```

Array and Object syntax work for every style prop in the theme specification, which means you can change most properties at a given breakpoint. Preferred way is to use array.

> To create custom breakpoints read [docs](https://chakra-ui.com/docs/features/responsive-styles#customizing-breakpoints).

Reviewer 3: I had some issue with this while our project `tsconfig` had `strict: false`. We should add a note that in order for everything to work as expected in `chakra-ui@1.3.2` and later you need to have `strict: true` in your `tsconfig` otherwise you will get a type error that `breakpoint.toString` are incompatible.

Here is the sandbox of the error: https://codesandbox.io/s/sleepy-sanne-t3fk4

> 

Reviewer 2: The `as` polymorphic prop

To create custom breakpoints read [docs](https://chakra-ui.com/docs/features/responsive-styles#customizing-breakpoints).

For more about responsive styles read [styled-system responsive styles docs](https://styled-system.com/responsive-styles/)

### Chakra Factory

Chakra factory serves as an object of chakra JSX elements, and also a function that can be used to enable custom component receive chakra's style props.

```tsx
import { chakra } from "@chakra-ui/react";

// object
<chakra.button
  sx={{
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  }}
>
  Click me
</chakra.button>;

// function
chakra("button", {
  baseStyle: {
    shadow: "lg",
    rounded: "lg",
    bg: "white",
  },
});
```

This reduces the need to create custom component wrappers and name them. Syntax is available for common html elements. See the reference for the full [list of elements](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/system.utils.ts#L9) supported.

> Chakra factory function can also that convert jsx element to chakra-enabled components so you can pass style props to them.

Reviewer 2: This is not clear enough. I don't understand what you want to say

Author: I removed this because we have something similar few lines above.

> For example `react-datepicker` can be wrapped in chakra factory function to be possible to pass style props to `<DatePicker />`

Reviewer 2: `to be possible` sounds weird... but I'm not an English expert :D

Example:

```tsx
import DatePicker from "react-datepicker";
import { chakra } from "@chakra-ui/react";

export const Datepicker = () => {

>   const ChakraDatepicker = chakra(DatePicker);

Reviewer 2: this should be moved outside the founction

  return <ChakraDatepicker sx={{ width: "100%", bg: "blue.100" }} />;
};
```

> Chakra factory function will pass `class` to `DatePicker input` component with all styles. In this case `input` will be rendered with custom `bg` and `width` styles.

Reviewer 2: `class` => `className`

This reduces the need to create custom component wrappers and name them. Syntax is available for common html elements. See the reference for the full [list of elements](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/system.utils.ts#L9) supported.

> For example `react-datepicker` can be wrapped in chakra factory function so we can pass `sx` style props to `<DatePicker />`

Reviewer 2: `sx` **or** style props
